[CI:DOCS] UPDATE MANPAGE_SYNTAX (commit,attach,auto-update) (#10526)
containers/podman: docs/source/markdown/podman-attach.1.md

**podman container attach** [*options*] *container*

## DESCRIPTION

> **podman attach** allows you to attach to a running *container* using the *container's name* or *ID*, either to view its ongoing output or to control it interactively.

Reviewer: I realize this isn't your change, but I bet @TomSweeneyRedHat is going to jump in here and say something about not using "you"...

Reviewer 2: indeed, 'you' is not good!
```suggestion
**podman attach** attaches to a running *container* using the *container's name* or *ID*, to either view its ongoing output or to control it interactively.
```

> You can detach from the *container* (and leave it running) using a configurable key sequence. The default

Reviewer 2: ```suggestion
The *container* can be detached (and leave it running) using a configurable key sequence. The default
```

sequence is `ctrl-p,ctrl-q`.
Configure the keys sequence using the **--detach-keys** option, or specifying

> it in the `containers.conf` file: see [**containers.conf(5)**](https://github.com/containers/common/blob/master/docs/containers.conf.5.md) for more information.

Reviewer 3: This link is showing up in the actual man page.

Author: Maybe we can filter them out with a regex before. In this case we could also leave the `**` in the `[]`. But sadly I am not familiar with regex.

Reviewer 3: We would need to do some sed magic.  @edsantiago @nalind Any ideas?

Reviewer: there's already some sed magic in the Makefile, for dealing with man pages. See `MANPAGES`, line 430. I'm not sure what needs to be done here but I'll take a look in the morning.

Author: It seems like links to other man pages or config files can only work in either markdown or in the man page. It would be really sad to dispense the idea of having commands linked to the correct .md files. But at the current stage until a solution is found for it, I will revert it.

Reviewer: If I understand the problem correctly, perhaps this could solve it:
```diff
diff --git a/Makefile b/Makefile
index 15d6d9fb6..70ba76e49 100644
--- a/Makefile
+++ b/Makefile
@@ -428,8 +428,10 @@ pkg/api/swagger.yaml: .gopathok
        make -C pkg/api

 $(MANPAGES): %: %.md .install.md2man docdir
-       @sed -e 's/\((podman[^)]*\.md)\)//g' -e 's/\[\(podman[^]]*\)\]/\1/g' \
-               -e 's;<\(/\)\?\(a\|a\s\+[^>]*\|sup\)>;;g' $<  | \
+       @sed -e 's/\((podman[^)]*\.md)\)//g' \
+            -e 's/\[\(podman[^]]*\)\]/\1/g' \
+            -e 's/\[\(\*[^]]*\)\](http[^)]\+)/\1/g' \
+            -e 's;<\(/\)\?\(a\|a\s\+[^>]*\|sup\)>;;g' $<  | \
        $(GOMD2MAN) -in /dev/stdin -out $(subst source/markdown,build/man,$@)

 .PHONY: docdir
```

Author: I meant, that the whole link should be filtered out in the manpage. In the way it is now it's just put in a less-than sign and greater-than sign. So everything that is between the less-than sign and greater-than sign should go.

The `[]` aren't a problem anymore. 

It would also be helpful to replace the `\` at the end of the line with two spaces. In this way the desired line break can be achieved.

Reviewer: >everything that is between the less-than sign and greater-than sign should go.

My patch converts `[**markup**](link)` to just `**markup**`, which isn't exactly the same thing, but has the same end effect: the `<link>` is gone from the man page:
```diff
diff -ur docs/build/man.ante/podman-attach.1 docs/build/man/podman-attach.1
--- docs/build/man.ante/podman-attach.1 2021-06-03 10:46:32.118576644 -0600
+++ docs/build/man/podman-attach.1      2021-06-03 13:50:10.629289923 -0600
@@ -19,8 +19,7 @@
 You can detach from the \fIcontainer\fP (and leave it running) using a configurable key sequence. The default
 sequence is \fB\fCctrl\-p,ctrl\-q\fR\&.
 Configure the keys sequence using the \fB\-\-detach\-keys\fP option, or specifying
-it in the \fB\fCcontainers.conf\fR file: see \fBcontainers.conf(5)\fP
-\[la]https://github.com/containers/common/blob/master/docs/containers.conf.5.md\[ra] for more information.
+it in the \fB\fCcontainers.conf\fR file: see \fBcontainers.conf(5)\fP for more information.

 .SH OPTIONS
 .SS \fB\-\-detach\-keys\fP=\fBsequence\fP
```
Is that not good enough?

Author: It works for http and https but no for relative links like `(podman.1.md)`. The link is gone, but the `[]` are still there, around the command.

Author: Quick overview for the visual errors for links in the manpages:

- `**[podman](podman.1.md)**` results in **podman**
 *for commands alone everything is working*

- `**[podman-generate-systemd --new](podman-generate-systemd.1.md#--new)**` results in 
**podman-generate-systemd --new**(podman-generate-systemd.1.md#--new)
*first possibility for command with an option*

- `[**podman-generate-systemd --new**](podman-generate-systemd.1.md#--new)` results in
 **podman-generate-systemd --new** <podman-generate-systemd.1.md#--new>
*second possibility for command with an option*
It seems like links that include an option don't work. I believe it is do to the space between the command and option.

- `**[containers.conf(5)](https://github.com/containers/common/blob/master/docs/containers.conf.5.md)**` results in
containers.conf(5) \<https://github.com/containers/common/blob/master/docs/containers.conf.5.md>
*first possibility for a link with http/https*

- `[**containers.conf(5)**](https://github.com/containers/common/blob/master/docs/containers.conf.5.md)` results in
**containers.conf(5)**
*second possibility for a link with http/https*
Since the `**` should be on the outside of the whole structure, it should be switched at the http/https links. 

Thank for the help with the regex. :)

Author: Can someone confirm, that the last `-e` line added to the makefile works for deleting the last two whitespaces in the `.md` file and creating a line break: 
```
$(MANPAGES): %: %.md .install.md2man docdir
	@sed -e 's/\((podman[^)]*\.md)\)//g' \
         -e 's/\[\(podman[^]]*\)\]/\1/g' \
         -e 's/\[\(\*[^]]*\)\](http[^)]\+)/\1/g' \
         -e 's;<\(/\)\?\(a\|a\s\+[^>]*\|sup\)>;;g' \
         -e 's/[[:blank:]]+/\n/g' $<  | \
	$(GOMD2MAN) -in /dev/stdin -out $(subst source/markdown,build/man,$@)
```

Reviewer 4: Just dropping in, but would `[[:blank:]]+` not match any sequence of two or more blank characters anywhere on any line?

Author: Yep i think so, but my first thought was also wrong. I wanted to replace the `\` with two spaces. And i believe this works:

```
$(MANPAGES): %: %.md .install.md2man docdir
	@sed -e 's/\((podman[^)]*\.md)\)//g' \
         -e 's/\[\(podman[^]]*\)\]/\1/g' \
         -e 's/\[\(\*[^]]*\)\](http[^)]\+)/\1/g' \
         -e 's;<\(/\)\?\(a\|a\s\+[^>]*\|sup\)>;;g' \
         -e 's/\\$//  /g' $<  | \
	$(GOMD2MAN) -in /dev/stdin -out $(subst source/markdown,build/man,$@)
```

Author: I edited the regex a little more. It should work as intended now.

## OPTIONS
#### **--detach-keys**=**sequence**

Specify the key **sequence** for detaching a *container*. Format is a single character `[a-Z]` or one or more `ctrl-<value>` characters where `<value>` is one of: `a-z`, `@`, `^`, `[`, `,` or `_`. Specifying "" will disable this feature. The default is `ctrl-p,ctrl-q`.

#### **--latest**, **-l** [=**false** | *true*]

Instead of providing the *container name* or *ID*, use the last created *container*. If you use methods other than Podman to run containers such as CRI-O, the last started *container* could be from either of those methods. The default is **false**.\
*IMPORTANT: This option is not available with the remote Podman client*

#### **--no-stdin** [=**false** | *true*]

> Do not attach STDIN. The default is **false**.

Reviewer: This isn't your change, but this is one of those cases that is nearly impossible for a human brain to read unambiguously. First, I don't believe that the default is false -- that makes no sense. Second, it really needs clarification. I think this should be something like "--no-stdin [=true|false" .... By default, **podman attach** includes stdin." But I think this needs clarification from other team members, because in my testing I can't actually get stdin to work in podman attach no matter what I do.

Author: I do believe that the default is false. Because it is --**no**-stdin and not just --stdin. https://github.com/containers/podman/blob/52dae693da0df1447b7f5210a4c842d5c5a8a401/cmd/podman/containers/attach.go#L52

I thought about how to express it alot, so I came up with this definition that is in the MANPAGE_SYNTAX:
If the OPTION is boolean, it is not requiered to set an argument. In this case the equal sign is put in brackets `[]` with the arguments, in difference to arguments that are not boolean as shown with **--answer**. The default argument that is shown for a boolean OPTION is given when the OPTION is not used in the command. If the OPTION is used in the command and no argument is specified, the OPTION will always become `true`.

Reviewer: Sorry, I wasn't clear. The ambiguity is in the meaning of "default": A reader of the man page, seeing the following documentation string:
```
    **--no-stdin** [=**false** | *true*]    [...] The default is **false**
```
...might have trouble understanding exactly how the flag operates:
* Does "default" apply to **the flag**? That is, if I, the end  user, specify `--no-stdin`, is that the same as `--no-stdin=false`, and hence, I need to explicitly specify `--no-stdin=true` to get the desired effect? Or,
* Does "default" apply to **podman itself**? That is, is the man page saying that `no-stdin` is false by default, hence podman does attach stdin? 
* 
Double negatives are evil. This is a horrible situation that needs clarification. IMO I would completely remove the `[=**false** | *true*]` string from the flag description, and remove "The default is false". They add clutter and confusion. Less is more:
```
   #### **--no-stdin**
   Do not attach STDIN.
```

Author: > Does "default" apply to podman itself? That is, is the man page saying that no-stdin is false by default, hence podman does attach stdin?

Yes this is what I mean. In the end it is optinial if you want to put an argument behind the option. This is why the whole thing is in `[]` including the equal sign. But since the default value of boolean options is sometimes `true` and sometimes `false` I thought it needed clarification. I had the choice to define what is default for me. 

> Does "default" apply to the flag? That is, if I, the end user, specify --no-stdin, is that the same as --no-stdin=false

No, thats is why I said in the MANPAGE_SYNTAX that if you type a option without putting anything behind it, it is always true. I believe this is the way in which it is currently working. 

> Double negatives are evil. This is a horrible situation that needs clarification.

That was my starting point and I think that to. But it not on me to change the option itself but in this case I would like to have it changed. 

>I would completely remove the [=**false** | *true*] string from the flag description, and remove "The default is false". They add clutter and confusion.

I though since there are these different cases of options that sometimes default to `true` and sometimes to `false` it should be mentioned. So you know what option will be in command automaticly if you dont specifiy them.

Reviewer: >No, thats is why I said in the MANPAGE_SYNTAX that if you type a option without putting anything behind it, it is always true. 

I interpret that as "Please read a long complicated document before you are allowed to understand our man pages".  I don't think that's fair to end users, who are simply trying to solve a problem by finding one small relevant piece in a large documentation set. I feel pretty strongly about removing the `[=false]` thing from the description, I think if you look at it from the perspective of a normal user, one without your background and expertise, you will agree that it is too confusing.

Author: You are completely right. For the normal user it is very confusing. I will remove it.

Just in general, is there no where this fits. Cause the only way currently for users to find out about it, is if they look into the code itself. I don't have a problem with that, but I have the feeling it should be mentioned somewhere. But you are right, maybe not in the manpages.

Author: @edsantiago @TomSweeneyRedHat @rhatdan @vrothberg 
What is the general opinion about the topic above. In the end, I find it really important. So it should be mentioned somewhere. I could live with leaving out the arguments `false` and `true` in the option line. But I would at least like to specifiy the default value in the description with a short sentence.
